src/Microsoft.Extensions.OData.Migration/TranslationMiddleware.cs

```diff
@@ -0,0 +1,123 @@
+﻿using Microsoft.AspNetCore.Http;
+using Microsoft.Data.Edm.Csdl;
+using Microsoft.Data.OData.Query;
+using Microsoft.Data.OData.Query.SemanticAst;
+using System;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Web;
+using System.Xml;
+
+namespace Microsoft.Extensions.OData.Migration
+{
+    /// <summary>
+    /// Translation Middleware currently converts V3 URI to V4 URI (future: converts query, request body as well)
+    /// </summary>
+    public class TranslationMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly MigrationOptions options;
+        private readonly Data.Edm.IEdmModel v3Model;
+        private readonly Microsoft.OData.Edm.IEdmModel v4Model;
+
+        /// <summary>
+        /// Instantiates a translation middleware.
+        /// </summary>
+        /// <param name="next">Delegate required by middleware</param>
+        /// <param name="options">Contains information necessary for segment translation</param>
+        public TranslationMiddleware(RequestDelegate next, MigrationOptions options)
+        {
+            this.options = options;
+            this._next = next;
+
+            using (XmlReader reader = XmlReader.Create(options.V3EdmxPath))
+            {
+                this.v3Model = EdmxReader.Parse(reader);
+            }

```

Comment:
per my other comment, I think the user should do this work. They might create the model using an ODataBuilder. We can document how to do this, but we probably don't need to assume how they get the model.